src/main/java/com/commercetools/sync/types/TypeSync.java

```diff
@@ -0,0 +1,270 @@
+            return true;
+        }
+
+        return false;
+    }
+
+    /**
+     * Given a set of type keys, fetches the corresponding types from CTP if they exist.
+     *
+     * @param keys the keys of the types that are wanted to be fetched.
+     * @return a {@link CompletionStage} which contains the list of types corresponding to the keys.
+     */
+    private CompletionStage<List<Type>> fetchExistingTypes(@Nonnull final Set<String> keys) {
+        return typeService
+                .fetchMatchingTypesByKeys(keys)
+                .exceptionally(exception -> {
+                    final String errorMessage = format(CTP_TYPE_FETCH_FAILED, keys);
+                    handleError(errorMessage, exception, keys.size());
+
+                    return emptyList();
+                });
+    }
+
+    /**
+     * Given a list of {@link Type} or {@link TypeDraft}, returns a map of keys to the
+     * {@link Type}/{@link TypeDraft} instances.
+     *
+     * @param types list of {@link Type}/{@link TypeDraft}
+     * @param <T>          a type that extends of {@link WithKey}.
+     * @return the map of keys to {@link Type}/{@link TypeDraft} instances.
+     */
+    private <T extends WithKey> Map<String, T> getKeysTypeMap(@Nonnull final List<T> types) {
+        return types.stream().collect(Collectors.toMap(WithKey::getKey, p -> p,
+            (typeA, typeB) -> typeB));
+    }
+
+    /**
+     * Given a {@link String} {@code errorMessage} and a {@link Throwable} {@code exception}, this method calls the
+     * optional error callback specified in the {@code syncOptions} and updates the {@code statistics} instance by
+     * incrementing the total number of failed types to sync.
+     *
+     * @param errorMessage The error message describing the reason(s) of failure.
+     * @param exception    The exception that called caused the failure, if any.
+     * @param failedTimes  The number of times that the failed types counter is incremented.
+     */
+    private void handleError(@Nonnull final String errorMessage, @Nullable final Throwable exception,
+                             final int failedTimes) {
+
+        syncOptions.applyErrorCallback(errorMessage, exception);
+        statistics.incrementFailed(failedTimes);
+    }
+
+    /**
+     * Given a list of type drafts, attempts to sync the drafts with the existing types in the CTP
+     * project. The type and the draft are considered to match if they have the same key.
+     *
+     * @param oldTypes old types.
+     * @param newTypes drafts that need to be synced.
+     * @return a {@link CompletionStage} which contains an empty result after execution of the update
+     */
+    private CompletionStage<TypeSyncStatistics> syncBatch(
+            @Nonnull final List<Type> oldTypes,
+            @Nonnull final List<TypeDraft> newTypes) {
+        final Map<String, Type> oldTypeMap = getKeysTypeMap(oldTypes);
+
+        return CompletableFuture.allOf(newTypes
+            .stream()
```

Comment:
Its always more readable if you indent the chan dots of the same level with the same indentation level.